File: cpp/epi/search_in_list.cc
#include <memory>

#include "../generic_types/list_node.h"
#include "generic_test.h"
using std::shared_ptr;

shared_ptr<ListNode<int>> SearchList(shared_ptr<ListNode<int>> L, int key) {
  while (L != nullptr && L->data != key) {
    L = L->next;
  }

  return L;
}

int SearchListWrapper(shared_ptr<ListNode<int>> L, int key) {
  auto result = SearchList(L, key);
  return result ? result->data : -1;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::string func_name = "SearchList";
  std::vector<std::string> param_names{"L", "key"};

  return GenericTestMain(args, "search_in_list.cc",
                         "../test_data/epi/search_in_list.tsv",
                         &SearchListWrapper, DefaultComparator{}, func_name, param_names);
}

File: java/epi/SearchInList.java
package epi;

import generic_types.ListNode;
import test_framework.EpiTest;
import test_framework.GenericTest;

public class SearchInList {

  public static ListNode<Integer> searchList(ListNode<Integer> L, int key) {
    return null;
  }

  @EpiTest(testDataFile = "../test_data/epi/search_in_list.tsv")
  public static int searchListWrapper(ListNode<Integer> L, int key) {
    ListNode<Integer> result = searchList(L, key);
    return result != null ? result.data : -1;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "SearchInList.java",
                new Object() {
                }.getClass().getEnclosingClass())
            .ordinal());
  }
}
